src/Npgsql/TypeMapping/ConnectorTypeMapper.cs

```diff
@@ -486,18 +486,69 @@ void BindRangeType(NpgsqlTypeHandler elementHandler, PostgresRangeType pgRangeTy
 
         #endregion Binding
 
-        internal NpgsqlDbType GetNpgsqlTypeByOid(uint oid)
+        internal (NpgsqlDbType npgsqlDbType, PostgresType postgresType, Type specificType) GetTypeInfoByOid(uint oid)
         {
-            if (!DatabaseInfo.ByOID.TryGetValue(oid, out var pgType))
+            if (!DatabaseInfo.ByOID.TryGetValue(oid, out var postgresType))
                 throw new InvalidOperationException($"Couldn't find PostgreSQL type with OID {oid}");
 
-            if (!Mappings.TryGetValue(pgType.Name, out var mapping) && !Mappings.TryGetValue(pgType.FullName, out mapping))
-                throw new InvalidOperationException($"No mapping found for PostgreSQL type {pgType.DisplayName}");
+            // Try to find the postgresType in the mappings 
+            if (TryGetTypeInfoFromMapping(postgresType, out NpgsqlDbType npgsqlDbType, out Type specificType))
+                return (npgsqlDbType, postgresType, specificType);
 
-            if (!mapping.NpgsqlDbType.HasValue)
-                throw new InvalidOperationException($"Invalid parameter type: {oid}");
+            if (postgresType is PostgresArrayType arrayType)
+            {
+                // Try to find the Elements' postgresType in the mappings 
+                if (TryGetTypeInfoFromMapping(arrayType.Element, out NpgsqlDbType elementType, out Type elementSpecificType))
+                    return (elementType | NpgsqlDbType.Array, postgresType, elementSpecificType);
+
+                // It might be an array of an unmapped enum type
+                if (arrayType.Element is PostgresEnumType)
+                    return (NpgsqlDbType.Array | NpgsqlDbType.Text, postgresType, null);
+            }
 
-            return mapping.NpgsqlDbType.Value;
+            // It might be an unmapped enum type
+            if (postgresType is PostgresEnumType)
+                return (NpgsqlDbType.Text, postgresType, null);
+
+            throw new InvalidOperationException($"No mapping found for PostgreSQL type {postgresType.DisplayName}");
+        }
+
+        private bool TryGetTypeInfoFromMapping(PostgresType pgType, out NpgsqlDbType type, out Type specificType)
+        {
+            if (Mappings.TryGetValue(pgType.Name, out var mapping) || Mappings.TryGetValue(pgType.FullName, out mapping))
+            {
+                if (mapping.NpgsqlDbType.HasValue)
+                {
+                    type = mapping.NpgsqlDbType.Value;
+                    specificType = null;
+                    return true;
+                }
+                var factoryType = mapping.TypeHandlerFactory.GetType();
```

Comment:
As written above, I don't think we should support deriving `SpecificType`, so most of this isn't needed. As it would make the whole method quite small some refactoring may be appropriate.